recipes-openxt/openxt/openxt-measuredlaunch/ml-functions

```diff
@@ -0,0 +1,375 @@
+parse_grub() {
+    local grub_conf=$1
+    exec 6< $grub_conf
+
+    local IFS=$'\n'
+    local lines=$(egrep -e "^[A-Z_]*=.*" $grub_conf)
+    for line in $lines; do
+        eval $line
+    done
+
+    local entry=$( grep "set default=" $grub_conf | cut -f2 -d= | tr -d \" )
+    [[ -z $entry ]] && entry=0
+
+    local count=0
+    while read -u 6 line; do
+        echo $line | grep -q menuentry || continue
+
+        if [[ $count -lt $entry ]]; then
+            count=$(( $count + 1 ))
+            continue
+        elif [[ $count -eq $entry ]]; then
+            break
+        elif [[ $count -gt $entry ]]; then
+            exec 6<&-
+            return 1;
+        fi
+    done
+
+    while read -u 6 line; do
+        echo $line | grep -q -e "^#" && continue
+        echo $line | grep -q -e "^}" && break
+
+        line=$(echo $line | sed -e 's/(/\\(/' -e 's/)/\\)/')
+        eval "echo $line" | awk '$1 ~ /multiboot|module/ { print substr($0, index($0,$2)) }'
+    done
+
+    exec 6<&-
+}
+
+# Function to calulate sinit hash and return the resulting hash
+# parameter 0: path to TXT AC module
+# return 0 on success
+# return 1 if passed invalid hashes
+hash_sinit() {
+    local acmod=$1
+    local policy="/tmp/.policy"
+    local tboot_idx='0x20000001'
+
+    lcp_readpol -i $tboot_idx -f $policy 2>&1 >/dev/null
+
+    tb_polgen --show $policy 2>&1 >/dev/null
+    [[ $? -ne 0 ]] && return 1
+
+    echo $(sinit_hash -p $policy $acmod|tr -d ' ')
+    rm $policy 2>&1 >/dev/null
+}
+
+# Function to perform a TPM hash extended
+# parameter 0: hash to extend
+# parameter 1: extending hash
+# return 0 on success
+# return 1 if passed invalid hashes 
+hash_extend() {
+    local first=$1
+    local second=$2
+    local buffer="/tmp/.buffer"
+
+    [[ $first == 0 ]] && first="0000000000000000000000000000000000000000"
+    [[ $(echo -n $first|wc -c) -ne 40 ]] && return 1
+    [[ $(echo -n $second|wc -c) -ne 40 ]] && return 1
+
+    echo -n $first | hex2bin >$buffer
+    echo -n $second | hex2bin >>$buffer
+
+    echo $(sha1sum $buffer | awk '{ print $1 }')
+    rm $buffer
+}
+
+# Function to calculate PCR 17
+# parameter 0: path to grub.cfg
+# return 0 on success
+# return 1 on failure
+pcr17() {
+    local root="${1:-/}"
+    local grub_conf=${2}
+    local existing=${3}
+    local pcr17=''
+
+    if [[ -n $existing ]]; then
+        local tpm_dir=$(find /sys/class -name tpm0)
+
+        [[ ! -d $tpm_dir ]] && return 1
+
+        pcr17=$(grep PCR-17 ${tpm_dir}/device/pcrs | cut -f2 -d: | tr -d ' ')
+
+        echo "${pcr17}"
+        return 0
+    fi
+
+    modules=$(parse_grub ${root}/$grub_conf)
+
+    if [[ $? -eq 0 ]]; then
+        local IFS=$'\n'
+        mod_num=0
+        for  modline in $modules; do
+            args=""
+            module=$(echo ${modline} | awk '{ print $1 }')
+            if [[ $(echo ${modline} | wc -w) -gt 2 ]]; then
+                cmdline=$(echo ${modline} | awk '{ print substr($0, index($0,$3)) }')
```

Comment:
In the current utilization, correct, but if it was possible to read the acm heap, then pcr17 can be calculated instead of re-used.